internal/auth/oauth_client.go

```diff
@@ -0,0 +1,354 @@
+		TokenEndpointAuthMethod: "private_key_jwt",
+		TokenEndpointAuthSigner: "ES256",
+		DpopBoundAccessTokens:   true,
+		Jwks:                    &jose.JSONWebKeySet{Keys: []jose.JSONWebKey{publicJwk}},
+	}
+}
+
+type AuthorizeState struct {
+	Verifier      string `json:"v"`
+	State         string `json:"s"`
+	TokenEndpoint string `json:"te"`
+}
+
+// Authorize implements OAuthClient.
+func (o *oauthClientImpl) Authorize(
+	dpopClient *DpopHttpClient,
+	i *identity.Identity,
+) (string, *AuthorizeState, error) {
+	pr, err := fetchOAuthProtectedResource(i)
+	if err != nil {
+		return "", nil, err
+	}
+
+	serverMetadata, err := fetchOauthAuthorizationServer(i, pr)
+	if err != nil {
+		return "", nil, err
+	}
+
+	verifier := oauth2.GenerateVerifier()
+
+	stateBytes := make([]byte, 12)
+	_, err = rand.Read(stateBytes)
+	if err != nil {
+		return "", nil, err
+	}
+	state := base64.URLEncoding.EncodeToString(stateBytes)
+
+	requestUri, err := o.makePushedAuthorizationRequest(
+		dpopClient,
+		i.DID.Identifier(),
+		serverMetadata,
+		state,
+		verifier,
+	)
+	if err != nil {
+		return "", nil, err
+	}
+
+	redirectUrl, err := url.Parse(serverMetadata.AuthEndpoint)
+	redirectUrl.RawQuery = url.Values{
+		"client_id":   {o.clientId},
+		"request_uri": {requestUri},
+	}.Encode()
+
+	return redirectUrl.String(), &AuthorizeState{
+		Verifier:      verifier,
+		State:         state,
+		TokenEndpoint: serverMetadata.TokenEndpoint,
+	}, nil
+}
+
+// ExchangeCode implements OAuthClient.
+func (o *oauthClientImpl) ExchangeCode(
+	dpopClient *DpopHttpClient,
+	code string,
+	issuer string,
+	state *AuthorizeState,
+) (string, error) {
+	clientAssertion, err := o.getClientAssertion(issuer)
+	if err != nil {
+		return "", err
+	}
+
+	log.Info().Msgf("toekn endpoint: %s", state.TokenEndpoint)
+	req, err := http.NewRequest(
+		http.MethodPost,
+		state.TokenEndpoint,
+		strings.NewReader(url.Values{
+			"client_id":     {o.clientId},
+			"grant_type":    {"authorization_code"},
+			"redirect_uri":  {o.redirectUri},
+			"code":          {code},
+			"code_verifier": {state.Verifier},
+
+			"client_assertion_type": {"urn:ietf:params:oauth:client-assertion-type:jwt-bearer"},
+			"client_assertion":      {clientAssertion},
+		}.Encode()),
+	)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := dpopClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		var errMsg json.RawMessage
+		json.NewDecoder(resp.Body).Decode(&errMsg)
+		return "", fmt.Errorf("failed to exchange code: %s", resp.Status, string(errMsg))
+	}
+
+	token, err := io.ReadAll(resp.Body)
+
+	return string(token), nil
+}
+
+type oauthProtectedResource struct {
+	AuthorizationServers []string `json:"authorization_servers"`
+}
+
+func fetchOAuthProtectedResource(i *identity.Identity) (*oauthProtectedResource, error) {
+	url, err := url.Parse(i.PDSEndpoint())
```

Comment:
Note that to my best knowledge this won't work with tailscale because the did document is not accessible on the tailnet domain name. My earlier workaround was to just manually pass in the PDS url since we assumed it would be on the Habitat node. This may not always be the case now that we're allowing external PDS